binders/kafka-binder/spring-cloud-stream-binder-kafka/src/test/java/org/springframework/cloud/stream/binder/kafka/KafkaBinderTests.java

```diff
@@ -4005,6 +4011,36 @@ void testObservationEnabledThroughProducerMessageHandlerCustomizer() throws Exce
 		setupBindingAndAssert("enable-observation.2", binder);
 	}
 
+
+	@Test
+	void testKafkaPartitionHandlerCalled() throws Exception {
+		Binder binder = getBinder();
+		ExtendedProducerProperties<KafkaProducerProperties> properties = createProducerProperties();
+		properties.setPartitionKeyExpression(
+			spelExpressionParser.parseExpression("headers['partitionKey']"));
+
+		DirectChannel outputChannel = createBindableChannel("output",
+			createProducerBindingProperties(createProducerProperties()));
+
+		Binding<MessageChannel> producerBinding = binder.bindProducer("foo",
+			outputChannel, properties);
+
+		KafkaMessageChannelBinder.ProducerConfigurationMessageHandler kafkaProducerMessageHandler =
+			(KafkaMessageChannelBinder.ProducerConfigurationMessageHandler) TestUtils.getPropertyValue(
+				producerBinding, "lifecycle", KafkaProducerMessageHandler.class);
+
+		assertThat(kafkaProducerMessageHandler.getKafkaPartitionHandler()).isNotNull();
```

Comment:
Do we need this assertion since if the partition handler is really null, the subsequent verifications on the spy will fail?

Reply:
It could be null in some usecases. Is there a rule or is it your personal preference?

Reply:
There is no rule. If this is possible to go null, sure, you can do the assertion.